File: Data Structures & Algorithms/string-matching-in-an-array/submission-0.py
class Solution:
    def stringMatching(self, words: List[str]) -> List[str]:
        
        def wordIsSubstringOfAnyString(word):
            for value in words:
                if word in value and word != value:
                    return True;
            return False;
        
        substrings = []

        for word in words:
            if wordIsSubstringOfAnyString(word):
                substrings.append(word);
        
        return substrings


        

File: Data Structures & Algorithms/word-break/submission-0.java
class Solution {
    Map<Integer, Boolean> cache = new HashMap<>();
    public boolean wordBreak(String s, List<String> wordDict) {
        return solve(s, wordDict, 0);
    }

    private boolean solve(String s, List<String> words, int i) {
        if (cache.containsKey(i))
            return cache.get(i);

        if (i == s.length())
            return true;
        
        boolean solvable = false;

        for (var word : words) {
            if (s.startsWith(word, i)) {
                solvable |= solve(s, words, i + word.length());
                if (solvable)
                    break;
            }
        }
        cache.put(i, solvable);
        return solvable;
    }
}
